---
title: "CAZy_analysis"
author: "Zeya Xue"
date: "3/29/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setting up packages and working directory 
```{r}
library(phyloseq);packageVersion("phyloseq")
library(ggplot2);packageVersion("ggplot2")
library(plyr)
library(dplyr)
library(superheat)
library(vegan)
library(reshape2)
library(forcats)
library(tidyverse)
library(ggpmisc) ## ggplot2 miscellaneous extension

path <- "~/Google Drive File Stream/My Drive/Lemay_lab/analysis/"
```

## Import the merged, normalized MEGARes output and make phyloseq object
```{r}
# Prep the count table
CountTab <- read.csv(file.path(path,"NovaSeq043/CAZy/CAZy_merged_gene_norm_tab.csv"), 
                     row.names = 1, header = TRUE, stringsAsFactors = FALSE)
#CountTab <- CountTab[ , !colnames(CountTab) %in% c("Class","Mechanism","Group","Gene")]
colnames(CountTab) <- gsub("X","",colnames(CountTab))
## format needed for importing into phyloseq
row.names(CountTab) <- CountTab$Gene
CountTab <- CountTab[,-1]
CountTab <- as.matrix.data.frame(CountTab)

# Read in metadata file
samdf <- read.table(file.path(path,"samdf/diet_data_only_average_summary_variables.txt"), 
                    header = TRUE)
#samdf <- samdf[ , !colnames(samdf) %in% c("X","X.1")]
rownames(samdf) <- samdf$subject_id

# merge as the phyloseq 
ps <- phyloseq(otu_table(CountTab, taxa_are_rows = TRUE), sample_data(samdf))
ps # 78241 taxa and 96 samples 
```

## Shannon diversity of genes
```{r}
a_div <- estimate_richness(ps, split = TRUE, measures = c("Shannon"))
a_div$subject_id <- row.names(a_div)

# merge a_div and samdf (right join)
a_df <- merge(x = samdf, y = a_div, by = "subject_id", all.y = TRUE)


a_df$subject_id <- factor(a_df$subject_id)
p1 <- ggplot(a_df, aes(x=subject_id, y=Shannon, fill=Future_use_authorization))+
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
p2 <- p1 + aes(x = fct_reorder(subject_id,Shannon))

# save to pdf 12x5
```

## Total number of genes
```{r}
# count the genes in each sample with non-zero values 
count_df <- colSums(CountTab!=0) %>% as.data.frame()
# rename the column header 
count_df <- rename(count_df, number = .) 
count_df$subject_id <- rownames(count_df)

# merge count_df and samdf (right join)
count_df <- merge(x = samdf, y = count_df, by = "subject_id", all.y = TRUE)

count_df$subject_id <- factor(count_df$subject_id)
p3 <- ggplot(count_df, aes(x=subject_id, y=number, fill=Future_use_authorization))+
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
p4 <- p3 + aes(x = fct_reorder(subject_id,number))

# save to pdf 12x5
```

## Parse based on CAZy families
```{r}
CountTab_split <- as.data.frame(CountTab)

# Split the gene and family name apart  
library(stringr)
# need extra backslashes due to special character
CountTab_split$cazy_fam <- str_split(as.character(row.names(CountTab)), "\\|", n=2, simplify = TRUE)[,2] 
# remove extra "|" at the end of the family names
CountTab_split$cazy_fam <- str_sub(CountTab_split$cazy_fam, end = str_length(CountTab_split$cazy_fam)-1)
# replace | with _
CountTab_split$cazy_fam <- gsub("\\|", "-", CountTab_split$cazy_fam)

# sum up all the abundace values based on family
CountTab_split_agg <- aggregate(.~cazy_fam, 
                                data=CountTab_split,
                                sum)
write.csv(CountTab_split_agg, file=file.path(path,"NovaSeq043/CAZy/CAZy_merged_family_norm_tab.csv"))

taxtab <- CountTab_split_agg$cazy_fam
taxtab <- as.matrix(taxtab)
row.names(taxtab) <- taxtab[,1]
colnames(taxtab) <- "family"

row.names(CountTab_split_agg) <- CountTab_split_agg$cazy_fam
CountTab_split_agg <- subset(CountTab_split_agg, select=-c(cazy_fam)) %>% as.matrix.data.frame()

# merge as a new phyloseq object
ps_fam <- phyloseq(otu_table(CountTab_split_agg, taxa_are_rows = TRUE),
                   tax_table(taxtab),sample_data(samdf))
ps_fam # 1507 taxa and 96 samples 
```

## Heatmaps based on CAZy familes 
```{r}
ps_fam_melt <- psmelt(ps_fam)
taxa.agg <- aggregate(Abundance ~ family + subject_id,
                      data = ps_fam_melt,
                      mean)
taxa.cast <- dcast(taxa.agg, family ~ subject_id, mean, value.var = "Abundance")
# need to change results from factor to numeric because of R
row.names(taxa.cast) <- taxa.cast$family
taxa.cast <- taxa.cast[, -1]

# if there is factor variable, make it numeric 
#indx <- sapply(taxa.cast, is.factor)
#taxa.cast[indx] <- lapply(taxa.cast[indx], function(x) as.numeric(as.character(x))) 

# calculate the sum of each family
taxa.cast30 <- cbind(taxa.cast, total = rowSums(taxa.cast)) 
taxa.cast30$taxa <- rownames(taxa.cast30)
# keep only the top 30 most abundant CAZy family
taxa.cast30 <- head(arrange(taxa.cast30,desc(total)), n = 30)
row.names(taxa.cast30) <- taxa.cast30$taxa
#remove the total and taxa columns
taxa.cast30 <- subset(taxa.cast30, select=-c(total, taxa)) 

# heatmap pf the most abundant 30 families
superheat(taxa.cast30,
          left.label.size = 0.1, 
          bottom.label.size = 0.1,
          bottom.label.text.size = 3, 
          left.label.text.size = 3,
          bottom.label.text.angle = 90,
          grid.hline = FALSE,
          n.clusters.cols=4, #K-means clustering
          bottom.label="variable")

# scree plot for determining clustering
## For 1 to 15 cluster centers
wssplot <- function(data, nc=15, seed=1234)
  {
    wss <- (nrow(data)-1)*sum(apply(data,2,var))
    for (i in 2:nc)
    {
      set.seed(seed)
      wss[i] <- sum(kmeans(data, centers=i)$withinss)
    }
    plot(1:nc, wss, type="b", xlab="Number of Clusters",
    ylab="Within groups sum of squares")
}

pdf(file.path(path,"NovaSeq043/CAZy/scree-top30-fam-kmeans.pdf"), w=5, h=5)
wssplot(taxa.cast30) %>% print()
dev.off()


#### Another method to make heatmap with ggplot2
# keep the top 30 taxa
taxa.agg30 <- subset(taxa.agg,taxa.agg$family %in% row.names(taxa.cast30))
# add the shannon div metric 
taxa.agg30 <- merge(x=taxa.agg30, y=a_div, on='subject_id')
# add the asa info
taxa.agg30 <- merge(x=taxa.agg30, y=samdf, on='subject_id')
# change subject_id column to factor
taxa.agg30$subject_id <- factor(taxa.agg30$subject_id)

p9 <- ggplot(taxa.agg30, aes(x=subject_id, y=family))+
  geom_raster(aes(fill=Abundance))+
  scale_fill_viridis_c()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
p10 <- p9 + aes(x = fct_reorder(subject_id, perKcal_fiber_tnfs))
# save to pdf 12*5
  
# heatmap of all families for an overview
taxa.agg$subject_id <- factor(taxa.agg$subject_id)
p5 <- ggplot(taxa.agg, aes(x=subject_id, y=family))+
  geom_raster(aes(fill=Abundance))+
  scale_fill_viridis_c()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# save to pdf
```


## find "most variable" families by range
```{r}
write.csv(CountTab_split, file.path(path, "NovaSeq043/CAZy/CAZy_family_norm_tab.csv"))

CountTab_split$cazy_fam <- factor(CountTab_split$cazy_fam)

# create an empty dataframe to hold results from loop 
cazy_fam_range <- data.frame(cazy_fam=factor(), min=numeric(), 
                             max=numeric(), range=numeric())

for (i in unique(CountTab_split$cazy_fam)){
  df <- subset(CountTab_split, cazy_fam %in% i) %>% 
    subset(select=-cazy_fam)
  rang <- range(df) 
  
  cazy_fam_range <- cazy_fam_range %>% add_row(cazy_fam=i, min=rang[1],
                                               max=rang[2], range=rang[2]-rang[1])
}

write.csv(cazy_fam_range, file.path(path, "NovaSeq043/CAZy/CAZy_family_range.csv"), row.names = FALSE)

# sort the range in descending order 
cazy_fam_range <- arrange(cazy_fam_range, desc(cazy_fam_range$range))
# keep the name of the first 30 families
taxa.agg_range <- subset(taxa.agg,taxa.agg$family %in% head(cazy_fam_range, 30)$cazy_fam)
# add the shannon div metric 
taxa.agg_range <- merge(x=taxa.agg_range, y=a_div, on='subject_id')

p6 <- ggplot(taxa.agg_range, aes(x=subject_id, y=family))+
  geom_raster(aes(fill=Abundance))+
  scale_fill_viridis_c()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
p7 <- p6 + aes(x = fct_reorder(subject_id, Shannon))

# save to pdf 12*5 
```

## find "most variable" families by stdev
```{r}
# create an empty dataframe to hold results from loop 
cazy_fam_stdev <- data.frame(cazy_fam=factor(), stdev=numeric())

for (i in unique(CountTab_split$cazy_fam)){
  df <- subset(CountTab_split, cazy_fam %in% i) %>% # select rpws that match i
    subset(select=-cazy_fam) # remove the column to keep only numbers 
  stdev <- sd(df %>% as.matrix()) # have to convert to numeric matrix for R to work
  
  cazy_fam_stdev <- cazy_fam_stdev %>% add_row(cazy_fam=i, stdev=stdev)
}

write.csv(cazy_fam_stdev, file.path(path, "NovaSeq043/CAZy/cazy_fam_stdev.csv"), row.names = FALSE)

# sort the range in descending order 
cazy_fam_stdev <- arrange(cazy_fam_stdev, desc(stdev))
# keep the name of the first 30 families
taxa.agg_stdev <- subset(taxa.agg,taxa.agg$family %in% head(cazy_fam_stdev, 30)$cazy_fam)
# add the shannon div metric 
taxa.agg_stdev <- merge(x=taxa.agg_stdev, y=a_div, on='subject_id')

p7 <- ggplot(taxa.agg_stdev, aes(x=subject_id, y=family))+
  geom_raster(aes(fill=Abundance))+
  scale_fill_viridis_c()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
p8 <- p7 + aes(x = fct_reorder(subject_id, Shannon))
```


## Correlation between CAZy and 16S diversity (scatter plot between alpha diversities)
```{r}
##### dot plot between alpha diversities
a_16S <- read.table(file.path(path, "qiime1_alphadiv.txt"))
a_16S$subject_id <- row.names(a_16S)
# rename column header to avoid confusion 
a_16S <- rename(a_16S, "Shannon_16S"="shannon")

# merge tables together by CAZy subject ids 
a_16S_cazy <- merge(x=a_16S, y=a_div, by="subject_id", all.y=TRUE)
a_16S_cazy <- merge(x=a_16S_cazy, y=subset(count_df, select=c("subject_id","number")))
# rename column header to avoid confusion 
a_16S_cazy <- rename(a_16S_cazy, "Shannon_CAZy"="Shannon")
a_16S_cazy <- rename(a_16S_cazy, "CAZy_gene_number"="number")
# melt dataframe 
a_16S_cazym <- melt(a_16S_cazy, id.vars=c("subject_id","Shannon_CAZy","CAZy_gene_number"),
                    variable.name="div_measure_16S", value.name="div")


pdf(file.path(path,"NovaSeq043/CAZy/correlation-CAZy_Shannon-diet_divs.pdf"), w=7, h=5)
ggplot(a_16S_cazym, aes(x=div, y=Shannon_CAZy))+
  geom_point()+ # avoid overplotting by geom_point()
  stat_smooth(method="lm", se=FALSE, formula=y~x)+
  facet_wrap(facets=~div_measure_16S, scales="free_x")+
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               formula = y~x, parse = TRUE, size = 3)+
     stat_fit_glance(aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
                   label.x = 0.3, label.y = "bottom", 
                   size = 3)
  #scale_color_viridis_d() # d for discrete, other methods are for continuous or binned data
dev.off()

pdf(file.path(path,"NovaSeq043/CAZy/correlation-CAZy_number-diet_divs.pdf"), w=7, h=5)
ggplot(a_16S_cazym, aes(x=div, y=CAZy_gene_number))+
  geom_point()+ # avoid overplotting by geom_point()
  stat_smooth(method="lm", se=FALSE, formula=y~x)+
  facet_wrap(facets=~div_measure_16S, scales="free_x")+
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               formula = y~x, parse = TRUE, size = 3)+
   stat_fit_glance(aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
                   label.x = 0.3, label.y = "bottom", 
                   size = 3)
  #scale_color_viridis_d() # d for discrete, other methods are for continuous or binned data
dev.off()
```

## Correlation between CAZy and QC-ed ASA24 data
```{r}
asa_df <- read.table(file.path(path,"samdf/diet_data_only_average_summary_variables.txt"), header=TRUE)

# merge asa data with cazy diversities
asa_cazy <- merge(x=asa_df, y=a_div, by="subject_id", all.y=TRUE)
asa_cazy <- merge(x=asa_cazy, y=subset(count_df, select=c("subject_id","number")), all.x=TRUE)

# change column names to be consistent with the previous figures
asa_cazy <- rename(asa_cazy, "Shannon_CAZy"="Shannon")
asa_cazy <- rename(asa_cazy, "CAZy_gene_number"="number")

# melt the data frame
asa_cazym <- melt(asa_cazy, id.vars=c("Shannon_CAZy","CAZy_gene_number","subject_id"),
                  measure.vars=c("perKcal_fiber_tnf","perKcal_fiber_tnfs","hei_asa24_totalscore"),
                  variable.name="asa_variable",
                  value.name="fiber_intake")
asa_cazym <- melt(asa_cazym, id.vars=c("asa_variable","fiber_intake","subject_id"),
                  variable.name = "CAZy_div_measure",
                  value.name = "div")

pdf(file.path(path,"NovaSeq043/CAZy/correlation-CAZy-perK_fiber.pdf"), w=6, h=4)
ggplot(asa_cazym, aes(x=fiber_intake, y=div))+
  geom_point()+ # avoid overplotting by geom_point()
  stat_smooth(method="lm", se=FALSE, formula=y~x)+
  # Use vars() to supply variables from the dataset
  facet_grid(rows=vars(CAZy_div_measure), cols=vars(asa_variable), scales="free")+ 
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               formula = y~x, parse = TRUE, size = 3)+
     stat_fit_glance(aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
                   label.x = 0.3, label.y = "bottom", 
                   size = 3)
dev.off()

pdf(file.path(path,"NovaSeq043/CAZy/correlation-CAZy-HEI.pdf"), w=4, h=4)
ggplot(asa_cazym, aes(x=fiber_intake, y=div))+
  geom_point()+ # avoid overplotting by geom_point()
  stat_smooth(method="lm", se=FALSE, formula=y~x)+
  # Use vars() to supply variables from the dataset
  facet_grid(rows=vars(CAZy_div_measure), cols=vars(asa_variable), scales="free")+ 
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               formula = y~x, parse = TRUE, size = 3)+
     stat_fit_glance(aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")),
                   label.x = 0.3, label.y = "bottom", 
                   size = 3)
dev.off()

```

# CAZy gene number vs seq depth
```{r}
reads <- read.table(file.path(path,"NovaSeq043/043_pipeline_libsize.tsv"), header = TRUE)

reads_cazy <- merge(reads, subset(count_df, select = c("number","subject_id")), 
                    by.x = "sampleID", by.y ="subject_id", 
                    all.x =  TRUE)

pdf(file.path(path,"NovaSeq043/CAZy/correlation_libsize_CAZy-genes.pdf"), w=4, h=3)
ggplot(reads_cazy, aes(x = flash, y = number))+
  geom_point()+
  geom_smooth(method='lm', se=FALSE)+ 
  theme_bw()+
  stat_poly_eq(formula = y ~ x,
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               parse = TRUE)+
  stat_fit_glance(method = 'lm',
                  method.args = list(formula = y ~ x),
                  geom = 'text',
                  label.x = "middle", label.y = "bottom", 
                  aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")))
dev.off()


reads_cazy2 <- merge(reads, a_div, 
                     by.x = "sampleID", by.y ="subject_id", 
                     all.x =  TRUE)

pdf(file.path(path,"NovaSeq043/CAZy/correlation_libsize_CAZy-shannon.pdf"), w=4, h=3)
ggplot(reads_cazy2, aes(x = flash, y = Shannon))+
  geom_point()+
  geom_smooth(method='lm', se=FALSE)+ 
  theme_bw()+
  stat_poly_eq(formula = y ~ x,
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.15,
               parse = TRUE)+
  stat_fit_glance(method = 'lm',
                  method.args = list(formula = y ~ x),
                  geom = 'text',
                  label.x = "middle", label.y = "bottom", 
                  aes(label = paste("P-value = ", signif(..p.value.., digits = 4), sep = "")))
dev.off()

```

